Source/GMCAbilitySystem/Public/Utility/GMASBoundQueue.h

```diff
@@ -0,0 +1,582 @@
+﻿// Fill out your copyright notice in the Description page of Project Settings.
+
+#pragma once
+
+#include "CoreMinimal.h"
+#include "GameplayTagContainer.h"
+#include "GMCMovementUtilityComponent.h"
+#include "InstancedStruct.h"
+#include "UObject/Object.h"
+#include "UObject/ConstructorHelpers.h"
+#include "GMASBoundQueue.generated.h"
+
+UENUM(BlueprintType)
+enum class EGMASBoundQueueOperationType : uint8
+{
+	None,
+	Add,
+	Remove,
+	Activate,
+	Cancel
+};
+
+USTRUCT(BlueprintType)
+struct FGMASBoundQueueOperationIdSet
+{
+	GENERATED_BODY()
+
+	UPROPERTY()
+	TArray<int> Ids = {};
+};
+
+USTRUCT(BlueprintType)
+struct GMCABILITYSYSTEM_API FGMASBoundQueueRPCHeader
+{
+	GENERATED_BODY()
+
+	UPROPERTY()
+	int32 OperationId { -1 };
+
+	UPROPERTY()
+	uint8 OperationTypeRaw { 0 };
+
+	EGMASBoundQueueOperationType GetOperationType() const
+	{
+		return static_cast<EGMASBoundQueueOperationType>(OperationTypeRaw);
+	}
+	
+	UPROPERTY()
+	FGameplayTag Tag { FGameplayTag::EmptyTag };
+
+	UPROPERTY()
+	FName ItemClassName { NAME_None };
+
+	UPROPERTY()
+	FGMASBoundQueueOperationIdSet PayloadIds {};
+
+	UPROPERTY()
+	FInstancedStruct InstancedPayload {};
+
+	UPROPERTY()
+	float RPCGracePeriodSeconds { 1.f };
+
+	UPROPERTY()
+	uint8 ExtraFlags { 0 };
+};
+
+template<typename C, typename T>
+struct GMCABILITYSYSTEM_API TGMASBoundQueueOperation
+{
+	FGMASBoundQueueRPCHeader Header {};
+	
+	// An actual class to be utilized with this, in case we need to instance it.
+	TSubclassOf<C> ItemClass { nullptr };
+
+	FInstancedStruct InstancedPayloadIds;
+	
+	// The struct payload for this item.
+	T Payload;	
+
+	// The instanced struct representation of this payload, used to actually
+	// bind for replication.
+	FInstancedStruct InstancedPayload;
```

Comment:
Keep in mind this is very unsafe as FInstancedStruct contains UObjects. The `UScriptStruct` of the wrapped struct, and the struct memory itself may contain `UPROPERTY` fields which the GC knows about through the `AddStructReferencedObjects` type trait.

I already talked to GRIM about implementing it for the GMC's bound variables since UObject* ones aren't referenced by the GC in some scenarios. You're meant to `UPROPERTY` instanced structs to prevent dangling pointers. 

If there is no way to make it UPROPERTY, there are other solutions:

- If this templated struct is held in a struct, you can implement the `WithAddStructReferencedObjects` type trait to manually add the instanced struct's references (and ItemClass as well) if they're not nullptr. 
- If this is held by a `UObject`, you can override `UObject::AddReferencedObjects()` and do the same.

I see this template is used by another template, so I would go up the chain until I find a UPROPERTY'd struct or UObject and handle it either way.

Reply:
It's worth noting that the payloads being used may have UObject* properties, but those are filled in on the side actually utilizing it, so I was not particularly concerned about dangling pointers in this case as opposed to an abstract way to sync the templated payload.

However, it's a valid concern if someone blithely used the bound queue for something else. I'm not going to be able to do a rewrite for the next couple of weeks, however, so if you'd like to see that change sooner, feel free to refactor and make a PR against the source branch!